framework/src/meshgenerators/MeshRepairGenerator.C

```diff
@@ -61,112 +63,136 @@ MeshRepairGenerator::generate()
+          const auto z_elem_node = elem_node(2);
+
+          if (MooseUtils::absoluteFuzzyEqual(x_node, x_elem_node, tol) &&
+              MooseUtils::absoluteFuzzyEqual(y_node, y_elem_node, tol) &&
+              MooseUtils::absoluteFuzzyEqual(z_node, z_elem_node, tol))
           {
-            const Real tol = _node_overlap_tol;
-            // Compares the coordinates
-            const auto x_node = (*node)(0);
-            const auto x_elem_node = elem_node(0);
-            const auto y_node = (*node)(1);
-            const auto y_elem_node = elem_node(1);
-            const auto z_node = (*node)(2);
-            const auto z_elem_node = elem_node(2);
-
-            if (MooseUtils::absoluteFuzzyEqual(x_node, x_elem_node, tol) &&
-                MooseUtils::absoluteFuzzyEqual(y_node, y_elem_node, tol) &&
-                MooseUtils::absoluteFuzzyEqual(z_node, z_elem_node, tol))
+            // Merging two nodes from the same element is almost never a good idea
+            if (elem->get_node_index(node) != libMesh::invalid_uint)
             {
-              // Coordinates are the same but it's not the same node
-              // Replace the node in the element
-              const_cast<Elem *>(elem)->set_node(elem->get_node_index(&elem_node)) = node;
-              nodes_removed.insert(elem_node.id());
-
-              _num_fixed_nodes++;
-              if (_num_fixed_nodes < 10)
-                _console << "Stitching a node at : " << *node << std::endl;
-              else if (_num_fixed_nodes == 10)
-                _console << "Node stitching will now proceed silently." << std::endl;
+              _console << "Two overlapping nodes in element " << elem->id() << " right by "
+                       << elem->vertex_average() << ".\n They will not be stitched" << std::endl;
+              continue;
             }
+
+            // Coordinates are the same but it's not the same node
+            // Replace the node in the element
+            const_cast<Elem *>(elem)->set_node(elem->get_node_index(&elem_node)) = node;
+            nodes_removed.insert(elem_node.id());
+
+            num_fixed_nodes++;
+            if (num_fixed_nodes < 10)
+              _console << "Stitching nodes " << *node << " and            " << elem_node
+                       << std::endl;
+            else if (num_fixed_nodes == 10)
+              _console << "Node stitching will now proceed silently." << std::endl;
           }
         }
       }
     }
-    _console << "Number of overlapping nodes which got merged: " << _num_fixed_nodes << std::endl;
-    if (mesh->allow_renumbering())
-      mesh->renumber_nodes_and_elements();
-    else
-    {
-      mesh->remove_orphaned_nodes();
-      mesh->update_parallel_id_counts();
-    }
   }
+  _console << "Number of overlapping nodes which got merged: " << num_fixed_nodes << std::endl;
+  if (mesh->allow_renumbering())
+  {
+    mesh->remove_orphaned_nodes();
+    mesh->renumber_nodes_and_elements();
+  }
```

Comment:
```suggestion
    mesh->renumber_nodes_and_elements();
```
Orphaned nodes are automatically removed in `renumber_nodes_and_elements`